Algorithm (stock-plugin): Enable PyTorch support to Image Robustness Toolkit + Refactor

jetkan-yk opened the pull request:
## Description

Enabled PyTorch support and streamlined workflow for Image Robustness Toolkit (blur, digital, environment, general).

## Motivation and Context

This update introduces PyTorch support and enhances the workflow for Image Corruption Toolbox by treating all models as complete pipelines.

Key changes include:

- **PyTorch Integration**: Users can now load any `torch.nn.Module` PyTorch model as a pipeline. 

- **Self-Contained Model Pipelines**: Preprocessing steps (e.g., resizing, cropping) are now expected to be included within the model file itself by defining a `predict` function that takes a batch of image paths as input and returns the prediction output. Example:
![image](https://github.com/user-attachments/assets/c442f391-abe0-4945-8cde-95977389d982)

- **Reduced End-to-End Runtime by 50%**: Streamlined testing workflow to reduce total runtime by half.
**Before**
![image](https://github.com/user-attachments/assets/0a1ec7d9-c2c4-4228-b66c-a737469e5d74)
**After**
![image](https://github.com/user-attachments/assets/0c681bd8-0ca7-4fa6-8421-852f6e8f7741)

- **Simplified Ground Truth Handling**: The `ground_truth_file` and `annotated_ground_truth_file` inputs have been merged into a single input to reduce complexity and improve user experience.

- **Workflow Refinements**: The `plugin_init`, `algo_init`, and `algo` functions have been refactored, making the initialization process more efficient and modular.

## Type of Change

<!-- Select the appropriate type of change: -->
- feat: A new feature
<!-- - fix: A bug fix -->
<!-- - chore: Routine tasks, maintenance, or tooling changes -->
<!-- - docs: Documentation updates -->
<!-- - style: Code style changes (e.g., formatting, indentation) -->
- refactor: Code refactoring without changes in functionality
<!-- - test: Adding or modifying tests -->
<!-- - perf: Performance improvements -->
<!-- - ci: Changes to the CI/CD configuration or scripts -->
<!-- - other: Other changes that don't fit into the above categories -->

## How to Test

1. Run **unit tests** and **end-to-end tests** with `pytest`.
2. Use **.vscode configurations** found in each corruption group’s directory to run a full pipeline test.

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [x]  I have tested the changes locally and verified that they work as expected.
- [x]  I have added or updated the necessary documentation (README, API docs, etc.).
- [x]  I have added appropriate unit tests or functional tests for the changes made.
- [x]  I have followed the project's coding conventions and style guidelines.
- [x]  I have rebased my branch onto the latest commit of the main branch.
- [x]  I have squashed or reorganized my commits into logical units.
- [x]  I have added any necessary dependencies or packages to the project's build configuration.
- [x]  I have performed a self-review of my own code.
- [x]  I have read, understood and agree to the Developer Certificate of Origin below, which this project utilises.

## Screenshots (if applicable)

### All tests passing

<img width="893" alt="image" src="https://github.com/user-attachments/assets/25b916ce-b87f-4fad-9e9e-f098c17d1828" />

## Additional Notes

[Add any additional information or context that might be relevant to reviewers.]

<details>
  <summary>Developer Certificate of Origin</summary>

 ```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
 ```
</details>

timlrx commented:
Thanks, the approach looks good to me. Just one thing to clarify - it seems with the refactored implementation, the image corruption will always run in pipeline mode and does not depend on the `run_pipeline` flag? @imda-peckyoke do you require this information in the API? If so we could probably just hard code this to true for all image corruption runs?

For the core module, could you clean it up by adding a test case to `aiverify-test-engine`  for the pytorch pipeline?

For the image corruption toolbox, the current e2e test covers only sklearn module. Could you add another test case using the pytorch model that you have created in the user defined files? There are also some unit tests that are failing due to the updates.

jetkan-yk commented:
> Thanks, the approach looks good to me. Just one thing to clarify - it seems with the refactored implementation, the image corruption will always run in pipeline mode and does not depend on the `run_pipeline` flag? @imda-peckyoke do you require this information in the API? If so we could probably just hard code this to true for all image corruption runs?

That works too. I'll see what @imda-peckyoke thinks.

> For the core module, could you clean it up by adding a test case to `aiverify-test-engine` for the pytorch pipeline?

Sure, I'll work on it.

> For the image corruption toolbox, the current e2e test covers only sklearn module. Could you add another test case using the pytorch model that you have created in the user defined files? There are also some unit tests that are failing due to the updates.

Yup, I'm wrapping up the unit tests. Will add pytorch e2e tests as well.

jetkan-yk commented:
# Updates

## 1. Added README for using PyTorch models

<img width="505" alt="image" src="https://github.com/user-attachments/assets/e17ac61e-7589-4497-9493-4ff61d86c7f1" />

## 2. Added `test_pytorchpipeline.py` core module tests

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/1b3e84f9-4cde-4736-840b-ed379983ded4" />

Note that the one test case is failing at the moment because it downloads the pipeline zip file from `https://github.com/aiverify-foundation/aiverify/raw/refs/heads/v2.x/aiverify-test-engine/tests/io/user_defined_files/pytorch_pipeline_files.zip`, which will only be available **after this PR has been merged** into branch `v2.x`.

## 3. Updated plugin e2e and unit tests

<img width="871" alt="image" src="https://github.com/user-attachments/assets/31dc0994-20c8-460f-8bb2-17245d233352" />

jetkan-yk commented:
> @jetkan-yk, there's a change in behaviour where the file paths to artifacts are now absolute system paths. We should keep to the previous behaviour where they are just relative paths with respect to the output folder i.e. something like `images/Horizontal_Motion_Blur/severity0/9.png`

Fixed with 8a688ff4ec09e9c3983e68745f654b98384172f2
<img width="479" alt="image" src="https://github.com/user-attachments/assets/6c990b2b-32f6-4249-95c1-e1cc549b7555" />

timlrx reviewed (approved):
Great, thanks for finally making AI Verify compatible with pytorch 🎉
